### DISCUSSION TITLE
[service plugin] Correct way to handle process termination

### DISCUSSION THREAD

**Discussion by asanikovich** [roadrunner-server/roadrunner#1627] [Q&A] [answered]
We run some processes by RR service plugin.
One of them has a long run logic and very important, which should not be broken (by termination)

In case of supervisor (we used before instead of RR service plugin), we can block SIGTERM signal (in place where important logic start) and increase timeout to wait before SIGKILL.
How to solve it with RR?

**Comment by rustatian** [comment]
Hey @asanikovich 👋🏻 
You may send any signal to the PHP process, stop it, kill, no limitations on that. If you don't have [this option](https://github.com/roadrunner-server/roadrunner/blob/master/.rr.yaml#L516), process will be finished. Otherwise it will be restarted.

**Reply by asanikovich** [reply]
I have option **remain_after_exit** enabled.
My case about graceful termination on Kubernetes new deploy.
In case of supervisor, supervisor send signals to child processes.

**Reply by rustatian** [reply]
RR sends `SIGINT` system call to the underlying PHP process (if you stop the whole RR or just terminate the service via RPC by its name). Then RR waits 5 seconds for the process to gracefully exit, and if the process still exists - sends `SIGKILL`.

**Reply by asanikovich** [reply]
1) In case of k8s send signal to pod, RR sends SIGINT to services process?
2) Can I modify default 5 seconds?

**Reply by rustatian** [reply, marked as answer]
1. As far as I remember, k8s sends a `SIGTERM` signal. When receiving this signal, RR will initiate a graceful stop, here is the RR timeout (not `service`): https://github.com/roadrunner-server/roadrunner/blob/master/.rr.yaml#L2169. And yeah, RR will then send `SIGINT` to the underlying services.
2. Unfortunately no, this timeout is fairly new, so still experimental. You may create a feature request for you previous request with the `service_name_in_logs` via RPC and this one in one FR.
